protocol/contracts/v2/TreasuryReservesVault.sol

```diff
@@ -0,0 +1,474 @@
+        _repay(msg.sender, strategy, repayAmount);
+    }
+
+    /**
+     * @notice A strategy calls to paydown all of it's debt
+     * This will pull the stables for the entire dUSD balance of the strategy, and burn the dUSD.
+     */
+    function repayAll(address strategy) external override returns (uint256 amountRepaid) {
+        amountRepaid = internalDebtToken.balanceOf(strategy);
+        _repay(msg.sender, strategy, amountRepaid);
+    }
+
+    function _repay(address _from, address _strategyAddr, uint256 _repayAmount) internal {
+        if (_repayAmount == 0) revert CommonEventsAndErrors.ExpectedNonZero();
+        if (globalRepaysPaused) revert RepaysPaused();
+
+        Strategy storage strategyData = strategies[_strategyAddr];
+        if (!strategyData.isEnabled) revert NotEnabled();
+        if (strategyData.repaysPaused) revert RepaysPaused();
+
+        emit Repay(_strategyAddr, _from, _repayAmount);
+
+        // Burn the dUSD tokens. This will fail if the repayment amount is greater than the balance.
+        uint256 burnedAmount = internalDebtToken.burn(_strategyAddr, _repayAmount);
+
+        // Pull the stables from the strategy.
+        stableToken.safeTransferFrom(_from, address(this), _repayAmount);
+
+        // If more stables are repaid than the total debt left, then that difference is a realised gain.
+        if (_repayAmount > burnedAmount) {
+            uint256 gain = _repayAmount-burnedAmount;
+            emit RealisedGain(_strategyAddr, gain);
+            totalRealisedGainOrLoss += int256(gain);
+        }
+    }
+
+    /**
+     * @notice The first step in a two-phase shutdown. Executor first sets whether a strategy is slated for shutdown.
+     * The strategy then needs to call shutdown as a separate call once ready.
+     */
+    function setStrategyIsShuttingDown(address strategy, bool isShuttingDown) external override onlyElevatedAccess {
+        Strategy storage strategyData = strategies[strategy];
+        if (!strategyData.isEnabled) revert NotEnabled();
+        emit StrategyIsShuttingDownSet(strategy, isShuttingDown);
+        strategyData.isShuttingDown = isShuttingDown;
+    }
+
+    /**
+     * @notice The second step in a two-phase shutdown. A strategy (automated) or executor (manual) calls
+     * to effect the shutdown. isShuttingDown must be true for the strategy first.
+     * The strategy executor is responsible for unwinding all it's positions first and sending stables to the TRV.
+     * All outstanding dUSD debt is burned, leaving a net gain/loss of equity for the shutdown strategy.
+     */
+    function shutdown(address strategyAddr) external override {
+        if (msg.sender != strategyAddr && !isElevatedAccess(msg.sender, msg.sig)) revert CommonEventsAndErrors.InvalidAccess();
+
+        Strategy storage strategyData = strategies[strategyAddr];
+        if (!strategyData.isEnabled) revert NotEnabled();
+        if (!strategyData.isShuttingDown) revert NotShuttingDown();
+
```

Comment:
by design, perhaps add a way to ensure strategy is unwinded before shutting down (in the event `automatedShutdown` isn't called)?
This could lead to a Denial of service (for Strategy) or possible inaccurate calculation (dashboard) if executors mistakenly execute `setShuttingDown` and `shutdown` immediately one after the next

Reply:
I don't think we can it 'by design'. For example if it's a Gnosis strategy, there's no easy way we can guarantee it's been fully unwound.

However I think this is going to be reasonably resilient:
1. The two steps means the TRV has to initiate it first
2. The shutdown() method will be more locked down - it will need a higher threshold of signers to call. So the idea is that it's thoroughly checked

Regardless of the shutdown, the AbstractStrategy still has a recoverToken() too, so can be pulled out after the fact

Reply:
Yeah I realized this after I reviewed the GnosisStrategy. 
Okay got it!